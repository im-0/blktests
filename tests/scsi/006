#!/bin/bash
# SPDX-License-Identifier: GPL-3.0+
# Copyright (C) 2018 Omar Sandoval
#
# Test that changing the cache type on a SCSI disk succeeds. Regression test
# for commit d772a65d8a6c ("Revert "scsi: core: avoid host-wide host_busy
# counter for scsi_mq""). Without the revert, this test will hang.

. tests/scsi/rc

DESCRIPTION="toggle SCSI cache type"

QUICK=1

device_requires() {
	_test_dev_is_scsi_disk
}

test_device() {
	echo "Running ${TEST_NAME}"

	local cache_types=(
		"write through"
		"none"
		"write back"
		"write back, no read (daft)"
	)
	local cache_type
	local original_cache_type
	local not_supported
	local new_cache_type

	original_cache_type="$(cat "${TEST_DEV_SYSFS}"/device/scsi_disk/*/cache_type)"
	for cache_type in "${cache_types[@]}"; do
		(
			echo "$cache_type" > "${TEST_DEV_SYSFS}"/device/scsi_disk/*/cache_type
		) |& awk 'inval=/Invalid argument$/ { ret=1 } !inval { print $0 } END { exit ret }'
		not_supported=$?

		if [ $not_supported -eq 0 ]; then
			new_cache_type="$(cat "${TEST_DEV_SYSFS}"/device/scsi_disk/*/cache_type)"
			if [ "${new_cache_type}" = "${cache_type}" ]; then
				echo "${cache_type}: success or not supported"
			else
				echo "${cache_type}: change failed"
			fi
		else
			echo "${cache_type}: success or not supported"
		fi
	done

	(
		echo "$original_cache_type" > "${TEST_DEV_SYSFS}"/device/scsi_disk/*/cache_type
	) |& grep -v "Invalid argument$"

	echo "Test complete"
}
