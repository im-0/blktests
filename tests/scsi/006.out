Running scsi/006
write through: success or not supported
none: success or not supported
write back: success or not supported
write back, no read (daft): success or not supported
Test complete
